roaring/src/bitmap/serialization.rs

```diff
@@ -47,6 +48,143 @@ impl RoaringBitmap {
+    #[inline]
+    pub fn from_bitmap_bytes(offset: u32, bytes: &[u8]) -> RoaringBitmap {
+        #[inline(always)]
+        fn next_multiple_of_u32(n: u32, multiple: u32) -> u32 {
+            (n + multiple - 1) / multiple * multiple
+        }
+        #[inline(always)]
+        fn next_multiple_of_usize(n: usize, multiple: usize) -> usize {
+            (n + multiple - 1) / multiple * multiple
+        }
+
+        const BITS_IN_ONE_CONTAINER: usize = u64::BITS as usize * BITMAP_LENGTH;
+        const BYTES_IN_ONE_CONTAINER: usize = BITS_IN_ONE_CONTAINER / 8;
+        assert_eq!(offset % 8, 0, "offset must be a multiple of 8");
+        let byte_offset = offset as usize / 8;
+        let n_containers_needed =
+            (bytes.len() + (BYTES_IN_ONE_CONTAINER - 1)) / BYTES_IN_ONE_CONTAINER + 1;
+        let mut containers = Vec::with_capacity(n_containers_needed);
+
+        let (offset, bytes) = if byte_offset % BYTES_IN_ONE_CONTAINER == 0 {
+            (offset, bytes)
+        } else {
+            let next_container_byte_offset =
+                next_multiple_of_usize(byte_offset, BYTES_IN_ONE_CONTAINER);
+
+            let number_of_bytes_in_first_container = next_container_byte_offset - byte_offset;
+            let number_of_bytes_copied_to_first_container =
+                bytes.len().min(number_of_bytes_in_first_container);
+
+            let (first_container_bytes, bytes_left) =
+                bytes.split_at(number_of_bytes_copied_to_first_container);
+            let (first_container_key, _) = util::split(offset);
+
+            let len: u64 = first_container_bytes.iter().map(|&b| b.count_ones() as u64).sum();
+            if len != 0 {
+                let mut bits: Box<[u64; BITMAP_LENGTH]> = Box::new([0; BITMAP_LENGTH]);
+                // Safety:
+                // * `first_container_bytes` is a slice of `bytes` and is guaranteed to be smaller than `BITMAP_LENGTH` u64s(= `BYTES_IN_ONE_CONTAINER` u8s)
+                // * `count` argument of `add` never equals to `BYTE_IN_ONE_CONTAINER` because at least one byte will be copied to the first container
+                //   This is guaranteed by the condition `byte_offset % BYTES_IN_ONE_CONTAINER != 0`
+                let bits_ptr = unsafe {
+                    bits.as_mut_ptr()
+                        .cast::<u8>()
+                        .add(BYTES_IN_ONE_CONTAINER - number_of_bytes_in_first_container)
+                };
+
+                debug_assert!(first_container_bytes.len() <= number_of_bytes_in_first_container);
+                // Safety:
+                // * `first_container_bytes` is a slice of `bytes` and is guaranteed to be smaller than or equal to `number_of_bytes_in_first_container`
+                unsafe {
+                    core::ptr::copy_nonoverlapping(
+                        first_container_bytes.as_ptr(),
+                        bits_ptr,
+                        first_container_bytes.len(),
+                    )
+                };
```

Comment:
Same below, but I don't think this is correct on Big Endian systems. E.g. I've got the bytes `[0x01, 0, 0, 0, 0, 0, 0, 0]`. If I plop those down into the `u64`s of the bitmap,  I've set the _least_ significant byte to 1 on a little endian system, but I've set the **most** significant byte to 1 on a big endian system. So the least significant bit of the u64 is set for a little endian system, but unset for a big endian system?

Reply:
I will add tests using miri, which is an interpreter of Rust's intermediate representation. That can even behave like big-endian system. I found the usage in [miri's README](https://github.com/rust-lang/miri?tab=readme-ov-file#cross-interpretation-running-for-different-targets).

I'll push the additional tests for big endian system and some patch (if needed) later (maybe 2 weeks later).

Reply:
I am not sure miri will detect algorithm correctness issues.

Reply:
At least, `miri` can emulate a big-endian system, allowing us to ensure that tests also pass on big-endian systems.